mypy/checker.py

```diff
@@ -1480,7 +1480,7 @@ def count_nested_types(self, typ: Instance, check_type: str) -> int:
     def visit_yield_stmt(self, s: YieldStmt) -> Type:
         return_type = self.return_types[-1]
         if isinstance(return_type, Instance):
-            if return_type.type.fullname() != 'typing.Iterator':
+            if not is_subtype(return_type, self.named_type('typing.Iterable')):
```

Comment:
OK, but then why is the code for visit_yield_from_stmt() right below not
making the same checks? The presence of `yield from` makes the function a
generator too. And yet this is legal:
```
def foo() -> Iterator[int]:
    yield 0
def bar() -> list[int]:
    yield from foo()
```